compositor_render/src/renderer/texture.rs

```diff
@@ -1,350 +1,184 @@
-                mapped_at_creation: false,
-                usage: wgpu::BufferUsages::COPY_DST | wgpu::BufferUsages::MAP_READ,
-                size: (Self::padded(resolution.width / 2) * resolution.height / 2) as u64,
-            }),
-        ];
+        let textures = YUVTextures::new(ctx, resolution);
+        let buffers = textures.new_download_buffers(ctx);
         Self {
-            yuv_textures: YUVTextures::new(ctx, resolution),
+            textures,
             buffers,
-            resolution: *resolution,
+            resolution: resolution.to_owned(),
         }
     }
 
-    /// extremely hacky workaround to download texture
-    /// TODO: do it properly, potentially we might want to use DownloadBuffer
-    pub fn download(&self, ctx: &WgpuCtx) -> YuvData {
-        self.transfer_content_to_buffers(&ctx.device, &ctx.queue);
-        let mut result = YuvData {
-            y_plane: Bytes::new(),
-            u_plane: Bytes::new(),
-            v_plane: Bytes::new(),
-        };
-        {
-            let size = self.yuv_textures.0[0].texture.size();
-            let buffer = BytesMut::with_capacity(size.width as usize * size.height as usize);
-
-            self.buffers[0]
-                .slice(..)
-                .map_async(MapMode::Read, |x| x.unwrap());
-            ctx.device.poll(wgpu::MaintainBase::Wait);
-
-            let range = self.buffers[0].slice(..).get_mapped_range();
-            let chunks = range.chunks(Self::padded(size.width as usize));
-            let mut buffer = buffer.writer();
-            for chunk in chunks {
-                buffer.write_all(&chunk[..size.width as usize]).unwrap();
-            }
-            result.y_plane = buffer.into_inner().into();
-        }
-        {
-            let size = self.yuv_textures.0[1].texture.size();
-            let buffer = BytesMut::with_capacity(size.width as usize * size.height as usize);
-
-            self.buffers[1]
-                .slice(..)
-                .map_async(MapMode::Read, |x| x.unwrap());
-            ctx.device.poll(wgpu::MaintainBase::Wait);
-
-            let range = self.buffers[1].slice(..).get_mapped_range();
-            let chunks = range.chunks(Self::padded(size.width as usize));
-            let mut buffer = buffer.writer();
-            for chunk in chunks {
-                buffer.write_all(&chunk[..size.width as usize]).unwrap();
-            }
-            result.u_plane = buffer.into_inner().into();
-        }
-        {
-            let size = self.yuv_textures.0[2].texture.size();
-            let buffer = BytesMut::with_capacity(size.width as usize * size.height as usize);
+    pub fn yuv_textures(&self) -> &YUVTextures {
+        &self.textures
+    }
 
-            self.buffers[2]
-                .slice(..)
-                .map_async(MapMode::Read, |x| x.unwrap());
-            ctx.device.poll(wgpu::MaintainBase::Wait);
+    pub fn resolution(&self) -> &Resolution {
+        &self.resolution
+    }
```

Comment:
```suggestion
    pub fn resolution(&self) -> Resolution {
        self.resolution
    }
```
`Resolution` is `Copy`